Reimplement pg_ctl start in python to simplify port opening detection

CyberDem0n opened the pull request.

Commit:
Reimplement pg_ctl start in python to simplify port open detection

ants commented on patroni/postgresql.py:
```diff
@@ -698,31 +690,28 @@ def start(self, block_callbacks=False):
         self.resolve_connection_addresses()
 
-        options = ' '.join("--{0}='{1}'".format(p, self._server_parameters[p]) for p, v in self.CMDLINE_OPTIONS.items()
+        options = ' '.join('--{0}="{1}"'.format(p, self._server_parameters[p]) for p, v in self.CMDLINE_OPTIONS.items()
```
I think server parameter values need to be escaped for shell here. Are there any sanity checks for server parameter keys in Patroni? If not, that might also have some security implications. Having write permissions for Patroni configuration will probably allow for arbitrary code execution no matter what we do, but it shouldn't be this easy :)

CyberDem0n replied:
https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L58

Only parameters defined in `CMDLINE_OPTIONS` will be passed via command line.
Also it contains function for checking values of parameters. So it should be safe.

Except cluster_name, listen_addresses and port...

ants commented on patroni/postgresql.py:
```diff
@@ -698,31 +690,28 @@ def start(self, block_callbacks=False):
+        pid = os.fork()
+        if pid == 0:
+            os.setsid()
+            os.execle("/bin/sh", "/bin/sh", "-c", cmd, {'PATH': os.environ.get('PATH')})
```
Not a Windows expert by far, but I'll hazard a guess that this doesn't work on Windows :) Do we care?

Before execing there is probably some cleanup stuff to do, I know that watchdog support wants that the watchdog device file handle is closed and not passed on to postgres. Probably good to close the rest too. Why did you decide to go through `/bin/sh` and `os.execle`? Using `subprocess.Popen` to execute postgres directly might be more portable and give us stuff like close_fds.

CyberDem0n replied:
Yeah it wont work on windows. Actually I took this example from original pg_ctl.c: https://github.com/postgres/postgres/blob/master/src/bin/pg_ctl/pg_ctl.c#L472

Will try to rewrite it with Popen...

Commit:
use subprocess.Popen instead of low-level functions

CyberDem0n commented:
Addressed code review. Any other comments?
Actually if you really want avoid  watchdog file handle being inherited by child processes and be always safe you should set `FD_CLOEXEC` flag on it,  we already doing so for REST API file handle: https://github.com/ants/patroni/blob/master/patroni/api.py#L427

ants commented:
Checking it out now, had to get a couple of things off my plate first. Good idea with `FD_CLOEXEC`, that addresses my main concern.
